### PR TITLE
Transaction tests for EIP-3860: Limit and meter initcode

### PR DESCRIPTION
**Pull request by gumb0**
These tests check transaction validity rules change of https://eips.ethereum.org/EIPS/eip-3860
Geth implementation in ethereum/go-ethereum#23847

### CODE REVIEW

File: src/TransactionTestsFiller/ttData/DataTestInitCodeTooBigFiller.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Creation transaction with initcode with size one bite above the max limit of EIP-3860
+DataTestInitCodeTooBig:
+    expectException:
+        London: ""
+        London+3860: "TR_InitCodeLimitExceeded"
```
Comment (gumb0) [inline src/TransactionTestsFiller/ttData/DataTestInitCodeTooBigFiller.yml]: This requires new rule in retesteth config:
```
      "TR_InitCodeLimitExceeded": "init code size limit exceeded"
```

**Comment by jochem-brouwer** [comment]
Is this one an updated version of #990?

**Comment by gumb0** [comment]
> Is this one an updated version of #990?

No, this is a follow-up to #990 (independent of it), here it's only _transaction tests_, testing specifically new rules for when transactions are invalid.

(Not that we have not updated EIP text yet to make transactions with too big initcode invalid, sorry.
But the tests here consider them invalid)

**Comment by gumb0** [comment]
> (Not that we have not updated EIP text yet to make transactions with too big initcode invalid, sorry. But the tests here consider them invalid)

EIP is updated now. https://github.com/ethereum/EIPs/pull/4757

**Comment by jochem-brouwer** [comment]
Ah yes this makes sense, we are not (yet) using the TransactionTests directory in JS, we should do that rather soon especially with these transactions. Thanks for the updates! :smile:

**Comment by jochem-brouwer** [comment]
Note: TransactionTests only supports LegacyTransactions ATM, so #990 should also include state tests or blockchain tests for situations where (1) initcode of Tx (legacy/2929/1559) is longer than max initcode and (2) sender does not pay the upfront cost of initcode analysis.

**Comment by gumb0** [comment]
> Note: TransactionTests only supports LegacyTransactions ATM, so #990 should also include state tests or blockchain tests for situations where (1) initcode of Tx (legacy/2929/1559) is longer than max initcode and (2) sender does not pay the upfront cost of initcode analysis.

It might be possible to add TransactionTests with different transaction types, I'll check.

But adding them as state tests is also a good idea.

**Comment by jochem-brouwer** [comment]
@gumb0 do these tests include tests regarding `CREATE/CREATE2` with sizes `maxInitCodeSize` and `maxInitCodeSize+1`?

EDIT: nvm, these are TxTests

File: src/TransactionTestsFiller/ttData/DataTestNotEnoughGasInitCodeFiller.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# Creation transaction with gas limit not enough to cover initcode charge of EIP-3860
+DataTestNotEnoughGasInitCode:
+    expectException:
+        London: ""
```
Comment (winsvega) [inline src/TransactionTestsFiller/ttData/DataTestNotEnoughGasInitCodeFiller.yml]: Is there an exception on London?

File: src/TransactionTestsFiller/ttData/DataTestInitCodeLimitFiller.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# Creation transaction with initcode with max size allowed by EIP-3860
+DataTestInitCodeLimit:
+    expectException: {}
```
Comment (winsvega) [inline src/TransactionTestsFiller/ttData/DataTestInitCodeLimitFiller.yml]: Is this field (expecException) required if its empty?

**Review by winsvega — CHANGES_REQUESTED** [review]
Update fork Merge+3860
Put EIP3860 in the test name or a new folder ttEIP3860
`London: ""`   in expectException can be deleted 
`expectException: {}` can be deleted

File: src/TransactionTestsFiller/ttData/DataTestInitCodeTooBigFiller.yml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# Creation transaction with initcode with size one bite above the max limit of EIP-3860
+DataTestInitCodeTooBig:
+    additionalForks:
+        - Merge+3860
+    expectException:
+        Merge: ""
+        Merge+3860: "TR_InitCodeLimitExceeded"
```
Comment (gumb0) [inline src/TransactionTestsFiller/ttData/DataTestInitCodeTooBigFiller.yml]: This requires a new mapping in retesteth config:
```
      "TR_InitCodeLimitExceeded" : "init code size limit exceeded",
```

**Comment by gumb0** [comment]
I would leave `Merge: ""` and `expectedException: {}` for explicitness, to show that we're testing that no exception is expected.

File: src/TransactionTestsFiller/ttEIP3860/DataTestNotEnoughGasInitCodeFiller.yml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# Creation transaction with gas limit not enough to cover initcode charge of EIP-3860
+DataTestNotEnoughGasInitCode:
+    expectException:
+        Merge: ""
+        Merge+3860: "TR_IntrinsicGas"
```
Comment (winsvega) [inline src/TransactionTestsFiller/ttEIP3860/DataTestNotEnoughGasInitCodeFiller.yml:7]: add Merge+3860 in `additionalForks` section that is missing here. the test is not generated for this fork.

File: src/TransactionTestsFiller/ttEIP3860/DataTestInitCodeLimitFiller.yml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# Creation transaction with initcode with max size allowed by EIP-3860
+DataTestInitCodeLimit:
+    expectException: {}
```
Comment (winsvega) [inline src/TransactionTestsFiller/ttEIP3860/DataTestInitCodeLimitFiller.yml:5]: additionalforks Merge+3860

File: src/TransactionTestsFiller/ttEIP3860/DataTestEnoughGasInitCodeFiller.yml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+# Creation transaction with gas limit just enough to cover initcode charge of EIP-3860
+DataTestEnoughGasInitCode:
+    expectException: {}
```
Comment (winsvega) [inline src/TransactionTestsFiller/ttEIP3860/DataTestEnoughGasInitCodeFiller.yml:5]: additional forks Merge+3860

**Review by winsvega — CHANGES_REQUESTED** [review]
by default tests are generated for each fork in progression of Forks. 
but since you use fork+eip here, need to specify in the test that it should also be generated using fork+eip by defining it in additionalForks section in the test. 

retesteth config change no needed. 

good to have small trasaction unit tests here. very clear.

File: src/TransactionTestsFiller/ttEIP3860/DataTestNotEnoughGasInitCodeFiller.yml [hunk 8]
```diff
@@ -0,0 +1,22 @@
+# Creation transaction with gas limit not enough to cover initcode charge of EIP-3860
+DataTestNotEnoughGasInitCode:
+    additionalForks:
+        - Merge+3860
+    expectException:
+        Merge: ""
```
Comment (winsvega) [inline src/TransactionTestsFiller/ttEIP3860/DataTestNotEnoughGasInitCodeFiller.yml:6]: if it is empty, then record not needed

**Pull request merged**